jlm/llvm/opt/ScalarEvolution.cpp

```diff
@@ -0,0 +1,209 @@
+    {
+      if (const auto thetaNode = dynamic_cast<const rvsdg::ThetaNode *>(structuralNode))
+        FindInductionVariables(thetaNode);
+      else
+      {
+        for (size_t r = 0; r < structuralNode->nsubregions(); r++)
+        {
+          TraverseRegion(structuralNode->subregion(r));
+        }
+      }
+    }
+  }
+}
+
+void
+ScalarEvolution::TraverseGraph(const rvsdg::Graph & rvsdg)
+{
+  rvsdg::Region & rootRegion = rvsdg.GetRootRegion();
+  TraverseRegion(&rootRegion);
+}
+
+void
+ScalarEvolution::FindInductionVariables(const rvsdg::ThetaNode * thetaNode)
+{
+  /*
+   * What do we have to do?
+   *
+   * Traverse the loop and find Basic Induction Variables (BIVs). These are variables with
+   * statements of the form i = i + c or i = i - c (increments by loop-invariant expressions), and
+   * put them in a set. Traverse the loop again and find General Induction Variables (GIVs)
+   * (linear combinations of BIVs) using the BIVs in the set. Put them in the set too.
+   *
+   * How can we traverse the loop?
+   * We can traverse top down, finding multiplication/addition nodes which use loop variables.
+   * Start by having all non loop-invariant loop variables in a set of induction variables, while
+   * traversing, look for addition nodes. If one of the loop variables is incremented by an
+   * expression that is not loop-invariant, remove it from the set.
+   * For other binary operations, remove them from the set for now.
+   *
+   * How do we make it robust?
+   * Need to handle different cases:
+   * 1. Loop variable and constant (loop variable at both rhs and lhs)
+   * 2. Loop variable and another loop variable
+   * 3. No loop variables
+   *
+   * How do we know if the expression is loop invariant?
+   * 1. It is a constant -> trivially invariant
+   * 2. It is a loop variable -> Check if it is loop invariant using ThetaLoopVarIsInvariant()
+   */
+
+  const std::vector<rvsdg::ThetaNode::LoopVar> loopVars = thetaNode->GetLoopVars();
+  InductionVariableSet inductionVariableCandidates{};
+
+  // Starting out, all non loop-invariant loop variables are induction variable candidates
+  for (auto loopVar : loopVars)
+  {
+    if (!rvsdg::ThetaLoopVarIsInvariant(loopVar))
+    {
+      // Only add loop-variant variables to the set, since loop-invariant variables are trivially
+      // not induction variables
+      inductionVariableCandidates.insert(loopVar.pre);
+    }
+  }
+  for (const auto node : rvsdg::TopDownTraverser(thetaNode->subregion()))
+  {
+    std::cout << node->DebugString() << '\n';
+    if (const auto simpleNode = dynamic_cast<rvsdg::SimpleNode *>(node))
+    {
+      if (!rvsdg::is<llvm::IntegerBinaryOperation>(simpleNode))
+      {
+        // TODO: Handle nodes that are not binary integer operations
+        continue;
+      }
+      // Check for basic induction variables (BIV)
+
+      assert(simpleNode->ninputs() == 2); // Assert because this should never happen
```

Comment:
Comment is useless. It does not provide more information than what the code already states.